# Programming for Data Analytics - Homework Four

Name : Samalie Piwan   
AndrewID : spiwan

In [91]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

#### PART A

**1) Download the dataset files, load them, and prepare the dataset for analysis**

Load the data into pandas dataframes

In [92]:
# Load all the file names into a list
filenames = []
file_directory = 'MachineLearningCVE'
for filename in os.listdir(file_directory):
    filenames.append(file_directory+"\\"+filename)

# Load the data in the files into a pandas dataframe
cids_df = pd.concat(map(pd.read_csv, filenames), ignore_index=True)
cids_df.head()

Destination Port   Flow Duration   Total Fwd Packets  \
0              54865               3                   2   
1              55054             109                   1   
2              55055              52                   1   
3              46236              34                   1   
4              54863               3                   2   

    Total Backward Packets  Total Length of Fwd Packets  \
0                        0                           12   
1                        1                            6   
2                        1                            6   
3                        1                            6   
4                        0                           12   

    Total Length of Bwd Packets   Fwd Packet Length Max  \
0                             0                       6   
1                             6                       6   
2                             6                       6   
3                             6                       6   
4                             0                       6   

    Fwd Packet Length Min   Fwd Packet Length Mean   Fwd Packet Length Std  \
0                       6                      6.0                     0.0   
1                       6                      6.0                     0.0   
2                       6                      6.0                     0.0   
3                       6                      6.0                     0.0   
4                       6                      6.0                     0.0   

   ...   min_seg_size_forward  Active Mean   Active Std   Active Max  \
0  ...                     20          0.0          0.0            0   
1  ...                     20          0.0          0.0            0   
2  ...                     20          0.0          0.0            0   
3  ...                     20          0.0          0.0            0   
4  ...                     20          0.0          0.0            0   

    Active Min  Idle Mean   Idle Std   Idle Max   Idle Min   Label  
0            0        0.0        0.0          0          0  BENIGN  
1            0        0.0        0.0          0          0  BENIGN  
2            0        0.0        0.0          0          0  BENIGN  
3            0        0.0        0.0          0          0  BENIGN  
4            0        0.0        0.0          0          0  BENIGN  

[5 rows x 79 columns]

**Data Exploration**

In [93]:
# Check the shape
cids_df.shape

(2830743, 79)

The resulting dataframe is quite large, with 2,830,743 rows and 79 columns

In [94]:
# Check the dataframe information
cids_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2830743 entries, 0 to 2830742
Data columns (total 79 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0    Destination Port             int64  
 1    Flow Duration                int64  
 2    Total Fwd Packets            int64  
 3    Total Backward Packets       int64  
 4   Total Length of Fwd Packets   int64  
 5    Total Length of Bwd Packets  int64  
 6    Fwd Packet Length Max        int64  
 7    Fwd Packet Length Min        int64  
 8    Fwd Packet Length Mean       float64
 9    Fwd Packet Length Std        float64
 10  Bwd Packet Length Max         int64  
 11   Bwd Packet Length Min        int64  
 12   Bwd Packet Length Mean       float64
 13   Bwd Packet Length Std        float64
 14  Flow Bytes/s                  float64
 15   Flow Packets/s               float64
 16   Flow IAT Mean                float64
 17   Flow IAT Std                 float64
 18   Flow IAT Max         

The dataframe contains mainly float and integer data, except the 'Label' column which contains an onject. This will be converted into a string

In [95]:
# Check for null values
null_values = cids_df.isnull().sum()
null_values[null_values > 0]

Flow Bytes/s    1358
dtype: int64

In [96]:
# Check the statistics of the data
cids_df.describe()

Destination Port   Flow Duration   Total Fwd Packets  \
count       2.830743e+06    2.830743e+06        2.830743e+06   
mean        8.071483e+03    1.478566e+07        9.361160e+00   
std         1.828363e+04    3.365374e+07        7.496728e+02   
min         0.000000e+00   -1.300000e+01        1.000000e+00   
25%         5.300000e+01    1.550000e+02        2.000000e+00   
50%         8.000000e+01    3.131600e+04        2.000000e+00   
75%         4.430000e+02    3.204828e+06        5.000000e+00   
max         6.553500e+04    1.200000e+08        2.197590e+05   

        Total Backward Packets  Total Length of Fwd Packets  \
count             2.830743e+06                 2.830743e+06   
mean              1.039377e+01                 5.493024e+02   
std               9.973883e+02                 9.993589e+03   
min               0.000000e+00                 0.000000e+00   
25%               1.000000e+00                 1.200000e+01   
50%               2.000000e+00                 6.200000e+01   
75%               4.000000e+00                 1.870000e+02   
max               2.919220e+05                 1.290000e+07   

        Total Length of Bwd Packets   Fwd Packet Length Max  \
count                  2.830743e+06            2.830743e+06   
mean                   1.616264e+04            2.075999e+02   
std                    2.263088e+06            7.171848e+02   
min                    0.000000e+00            0.000000e+00   
25%                    0.000000e+00            6.000000e+00   
50%                    1.230000e+02            3.700000e+01   
75%                    4.820000e+02            8.100000e+01   
max                    6.554530e+08            2.482000e+04   

        Fwd Packet Length Min   Fwd Packet Length Mean  \
count            2.830743e+06             2.830743e+06   
mean             1.871366e+01             5.820194e+01   
std              6.033935e+01             1.860912e+02   
min              0.000000e+00             0.000000e+00   
25%              0.000000e+00             6.000000e+00   
50%              2.000000e+00             3.400000e+01   
75%              3.600000e+01             5.000000e+01   
max              2.325000e+03             5.940857e+03   

        Fwd Packet Length Std  ...   act_data_pkt_fwd   min_seg_size_forward  \
count            2.830743e+06  ...       2.830743e+06           2.830743e+06   
mean             6.891013e+01  ...       5.418218e+00          -2.741688e+03   
std              2.811871e+02  ...       6.364257e+02           1.084989e+06   
min              0.000000e+00  ...       0.000000e+00          -5.368707e+08   
25%              0.000000e+00  ...       0.000000e+00           2.000000e+01   
50%              0.000000e+00  ...       1.000000e+00           2.400000e+01   
75%              2.616295e+01  ...       2.000000e+00           3.200000e+01   
max              7.125597e+03  ...       2.135570e+05           1.380000e+02   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  2.830743e+06  2.830743e+06  2.830743e+06  2.830743e+06  2.830743e+06   
mean   8.155132e+04  4.113412e+04  1.531825e+05  5.829582e+04  8.316037e+06   
std    6.485999e+05  3.933815e+05  1.025825e+06  5.770923e+05  2.363008e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.100000e+08  7.420000e+07  1.100000e+08  1.100000e+08  1.200000e+08   

           Idle Std      Idle Max      Idle Min  
count  2.830743e+06  2.830743e+06  2.830743e+06  
mean   5.038439e+05  8.695752e+06  7.920031e+06  
std    4.602984e+06  2.436689e+07  2.336342e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.00

From the data description, the data generally has data within a wide range of values. Some columns have data within thousands while others have data within tens and ones. This will need to be scaled to normalize the values

In [97]:
# Check for the details of unqiue labels
cids_df[' Label'].unique()

array(['BENIGN', 'DDoS', 'PortScan', 'Bot', 'Infiltration',
       'Web Attack � Brute Force', 'Web Attack � XSS',
       'Web Attack � Sql Injection', 'FTP-Patator', 'SSH-Patator',
       'DoS slowloris', 'DoS Slowhttptest', 'DoS Hulk', 'DoS GoldenEye',
       'Heartbleed'], dtype=object)

In [98]:
# Check for the count of unqiue labels
cids_df[' Label'].nunique()

15

From the above, we can see that there are 15 unique types intrusion security attacks

In [99]:
# Check for columns where data has no variance i.e the values is always 0
var_values = cids_df.var(numeric_only = True)
zero_var_values = var_values[var_values == 0]
zero_var_values

 Bwd PSH Flags           0.0
 Bwd URG Flags           0.0
Fwd Avg Bytes/Bulk       0.0
 Fwd Avg Packets/Bulk    0.0
 Fwd Avg Bulk Rate       0.0
 Bwd Avg Bytes/Bulk      0.0
 Bwd Avg Packets/Bulk    0.0
Bwd Avg Bulk Rate        0.0
dtype: float64

We observe that there are 8 columns with variance as 0, meaning they have all values as 0. These can be dropped during data cleaning, as the values are always constant and we do not gain information from them.

**Data Cleaning and Transformation**

Handle the null values in the 'Flow Bytes/s' column using imputation with the mean

In [100]:
cids_df['Flow Bytes/s'] = cids_df['Flow Bytes/s'].fillna(cids_df['Flow Bytes/s'].mean())

# Check if NaN values still exist
null_values_clean = cids_df.isnull().sum()
null_values_clean[null_values_clean > 0]

Series([], dtype: int64)

Handle the columns with no variance

In [101]:
# Check the number of columns before dropping the data
len(cids_df.columns)

79

In [102]:
zero_var_cols = list(zero_var_values.index)
cids_df = cids_df.drop(columns = zero_var_cols)

# Check the number of columns after dropping the data
len(cids_df.columns)

71